### ISSUE TITLE
OpenSSL should disallow custom elliptic curves in certificate chains.

### ISSUE THREAD

**Issue by Immortalem** [openssl/openssl#12139]
### Affected Versions
1.1.1g, 1.0.2u

### Setup
OpenSSL has been compiled from source in an alpine docker container. The following excerpt of the used docker file shows the build related commands:
```
RUN ./config --prefix=/build/ --openssldir=/build/ no-async 
RUN make -s && make install_sw -s
```
After building it the respective OpenSSL version and required libraries have been copied into an minimal docker container (from scratch).

The server is later started using the following command:
`openssl s_server -accept 4433 -Verify 100 -cert /cert/inputCerts/rootv3.pem -key /cert/keys/rootv3.pem -CAfile /cert/inputCerts/root.pem -verify_return_error`

All files mentioned in this report (certificates, keys, etc.) have been included in the attached zip archive.

### Issue
OpenSSL allows certificate chains in which custom elliptic curves are used. The recent Curveball vulnerability in Windows leads to the question if this should be allows. I think to prevent any mishaps from occurring this shouldn't be allowed by default. As far as I'm aware any public certificate chains use named curves so this *shouldn't* cause any incompatibilities.

### Reproduction
Uses OpenSSL 1.1.1 client on Ubuntu 18.04.
Connect to the started server using the following command:
`openssl s_client -connect localhost:4433 -cert ECROOTv3_CAv3CustomCurve_LEAF_ECv3__leaf_certificate1.pem -key eckey_secp256r1_3.pem -CAfile ECROOTv3_CAv3CustomCurve_LEAF_ECv3__ca_certificate1.pem`

### Expected Result
OpenSSL should reject the certificate chain since custom elliptic curves bear more security concerns than gains.

### Actual Result
OpenSSL happily accepts the certificate and proceeds with the handshake.

### Attachment
[opensslCustomEC.zip](https://github.com/openssl/openssl/files/4774643/opensslCustomEC.zip)

**Comment by romen** [comment]
Would disabling this by default in 1.1.1 create a risk for people that built their own CA and PKI using openssl versions where the default encoding was with explicit parameters?

If the above risk is considerable, I think a version of #12683 is desirable in 1.1.1, but maybe as an opt in?

**Comment by t8m** [comment]
This is clear violation of RFC5280 but yes, for 1.1.1 we need to consider the conditions under which the check is enabled.

**Comment by kroeckx** [comment]
I don't really have a check for that currently. I have a list of
allowed nids, those allowed by the BRs: NID_X9_62_prime256v1,
NID_secp384r1, NID_secp521r1

I think that check would trigger such certs, but I'm not sure. But
it will clearly trigger on certs using brainpool.

If you have an example certs and tell me how to detect it, I'd be
happy to make x509lint check that, and run it over the database.

**Comment by kroeckx** [comment]
I guess I should just have looked at your PR and used the
EC_GROUP_get_asn1_flag()

**Comment by kroeckx** [comment]
Or reading it again, that might also not work?

**Comment by kroeckx** [comment]
I think I can see it using X509_PUBKEY_get0_param ...

**Comment by kroeckx** [comment]
> @kroeckx could you please check your cert database if you see there any intermediates with explicitly encoded ECDSA parameters?

x509lint has been updated, processing the databasa, so far nothing
found.

**Comment by kroeckx** [comment]
So I found some, but so far they're all weird.

**Comment by t8m** [comment]
The EC_GROUP_get_asn1_flag() works, although accidentally only.

**Comment by t8m** [comment]
Let's reopen as we want to fix it also on 1.1.1 branch.

**Comment by mattcaswell** [comment]
This is now also fixed in 1.1.1, so closing.

**Comment by charlesmunger** [comment]
> Would disabling this by default in 1.1.1 create a risk for people that built their own CA and PKI using openssl versions where the default encoding was with explicit parameters?
> 
> If the above risk is considerable, I think a version of #12683 is desirable in 1.1.1, but maybe as an opt in?

This did happen to my home setup - I updated and started getting this error out of nginx. For others who come to this issue, specify `-param_enc named_curve` when generating with `openssl ecparam`.
